Dispatcher/UI/TextFields/TextFieldWithValidation.swift

```diff
@@ -0,0 +1,149 @@
+        switch type {
+        case .email:
+            setEmailTextField()
+        case .password:
+            setPasswordTextField()
+        case .reEnterPssword:
+            setReEnterPasswordTextField()
+        }
+    }
+    
+    private func setEmailTextField() {
+        self.styleTextFieldPlaceHolder(placeholderText: TextCostants.emailTextFieldPlaceholder, fontColor: UIColor(named: ColorsPalleteNames.labelsTextColor))
+        self.textContentType = .emailAddress
+    }
+    
+    private func setPasswordTextField() {
+        self.styleTextFieldPlaceHolder(placeholderText: TextCostants.passwordTextFieldPlaceholder, fontColor: UIColor(named: ColorsPalleteNames.labelsTextColor))
+        self.isSecureTextEntry = true
+        setIsSecurePasswordIcon()
+    }
+    
+    private func setReEnterPasswordTextField() {
+        self.styleTextFieldPlaceHolder(placeholderText: TextCostants.reEnterPasswordTextFieldPlaceholder, fontColor: UIColor(named: ColorsPalleteNames.labelsTextColor))
+        self.isSecureTextEntry = true
+        setIsSecurePasswordIcon()
+    }
+
+    private func setIsSecurePasswordIcon() {
+        let image = (self.isSecureTextEntry ? UIImage(named: "hideText") : UIImage(named: "showText")) ?? UIImage()
+        
+        let showOrHideTextIcon = UIButton()
+        showOrHideTextIcon.setImage(image, for: .normal)
+        showOrHideTextIcon.imageView?.contentMode = .scaleAspectFit
+        showOrHideTextIcon.addTarget(self, action: #selector(showOrHidePasswordPressed) , for: .touchUpInside)
+        self.addIconToTextField(position: .end, icon: showOrHideTextIcon)
+    }
+    
+    func styleValidTextField() {
+        self.textColor = UIColor(named: ColorsPalleteNames.labelsTextColor)
+        self.layer.borderColor = UIColor(named: ColorsPalleteNames.labelsTextColor)?.cgColor
+        self.layer.borderWidth = 0.5
+    }
+    
+    func styleErrorTextField() {
+        self.textColor = .red
+        self.styleTextFieldPlaceHolder(placeholderText: self.placeholder ?? "", fontColor: .red)
+        self.layer.borderColor = UIColor.red.cgColor
+        self.layer.borderWidth = 0.5
+    }
+    
+    @objc private func showOrHidePasswordPressed() {
+        self.isSecureTextEntry = !self.isSecureTextEntry
+        setIsSecurePasswordIcon()
+    }
+
+    func textFieldDidBeginEditing(_ textField: UITextField) {
+        textField.placeholder = ""
+    }
+
+    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
+        self.resignFirstResponder()
+        switch type {
+        case .email:
+            validateEmail(userInput: self.text)
+        case .password:
+            validatePassword(userInput: self.text)
+        case .reEnterPssword:
+            comparePassword(userInput: self.text)
+        }
+           return true
+    }
+    
+    //MARK: Input validation funcs
+
+    private func validateEmail(userInput: String?) {
+       let emailRegex = "[A-Z0-9a-z._%+-]+@[A-Za-z0-9.-]+"
```

Comment:
code is similar between the validateEmail and validatePassword functions.
Better extract the regex check into a dedicated function in the Util file and just send inside the relevant parameters and a completion